allennlp/models/encoder_decoders/nlvr_semantic_parser.py

```diff
@@ -0,0 +1,365 @@
+
+    @classmethod
+    def _compute_new_states(cls,  # type: ignore
+                            state: NlvrDecoderState,
+                            log_probs: torch.Tensor,
+                            hidden_state: torch.Tensor,
+                            memory_cell: torch.Tensor,
+                            action_embeddings: torch.Tensor,
+                            attended_question: torch.Tensor,
+                            considered_actions: List[List[int]],
+                            allowed_actions: List[Set[int]],
+                            max_actions: int = None) -> List[NlvrDecoderState]:
+        """
+        We override the ``_compute_new_states`` method here to compute new states based on coverage
+        loss.  For each action in considered actions, if it is in the agenda, we update the checklist
+        and recalculate the score. We'll have to iterate through considered actions in either case
+        because we have to update action histories anyway.
+        """
+        # TODO(pradeep): We do not have a notion of ``allowed_actions`` for NLVR for now, but this
+        # may be applicable in the future.
+        # Note that we're not sorting these probs. We'll score checklists and sort states based on
+        # checklist scores later.
+        probs = torch.exp(log_probs)
+        # batch_index -> [(group_index, action_index, action, checklist, score)]
+        next_states_info: Dict[int, List[Tuple[int, int, int, Variable, Variable]]] = defaultdict(list)
+        for group_index, (batch_index, instance_action_probs) in enumerate(zip(state.batch_indices, probs)):
+            instance_agenda = state.agenda[group_index]
+            instance_agenda_mask = state.agenda_mask[group_index]
+            instance_checklist = state.checklist[group_index]
+            # action_prob is a Variable.
+            for action_index, action_prob in enumerate(instance_action_probs):
+                # This is the actual index of the action from the original list of actions.
+                action = considered_actions[group_index][action_index]
+                if action == -1:
+                    # Ignoring padding.
+                    continue
+                # If action is not in instance_agenda, mask_variable, and checklist_addition will be
+                # all 0s.
+                checklist_mask = instance_agenda.eq(action)
+                checklist_addition = checklist_mask.float() * action_prob
+                new_checklist = instance_checklist + checklist_addition
+                new_score = cls.score_instance_checklist(new_checklist, instance_agenda_mask)
+                next_states_info[batch_index].append((group_index, action_index, action,
+                                                      new_checklist, new_score))
+        new_states = []
+        for batch_index, states_info in next_states_info.items():
+            sorted_states_info = sorted(states_info, key=lambda x: x[-1].data.cpu().numpy()[0],
```

Comment:
Calling `.data.cpu().numpy()` in the key function is going to be _incredibly_ slow.  You want to do a single batched `.cpu()` operation.

Reply:
Fixed.